proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,278 @@
+initaited by the account authority, and a special instruction by which fee payer accepts
+amount of lamports they are willing to pay as application fees per account.
+
+#### PayApplicationFees Instruction
+
+With this instruction, the fee payer accepts to pay application fees specifying the amount.
+This instruction **MUST** be included in the transaction that interacts with dapps having application fees.
+This instruction is like an irrevokable transfer if the payer has enough funds
+i.e even if the transaction fails the payer will end up paying.
+If this instruction is added then even if the dapp does not check the application fees the payer ends
+up paying.
+If the payer does not have enough balance the transaction fails with error `InsufficientFunds`.
+
+It requires:
+
+Accounts :
+* List of accounts that need an application fees.
+
+Argument : Corresponding list of application fees for each account in `Accounts`.
+
+The arguments list must be of same length as number of accounts.
+The index of fees and account should match.
+The account fees for each account is 8 bytes represented in rust type `u64`
+
+#### CheckApplicationFees Instruction
+
+This instruction will check if an application fee for an account is paid.
+It requires :
+
+* Account where fees are paid.
+
+Argument: required fees in lamport (u64).
+
+If application fees are not paid or are paid insufficiently this instruction will return
+an error. The idea is dapp developer uses this instruction to check if the required fees
+are paid and fail the transaction if they are not paid or partially paid.
+In case of partial payment, the user will lose the partially paid amount.
+A payer may overpay for the fees. This instruction can be called multiple times across multiple instructions.
+
+#### Rebate Instruction
+
+This instruction should be called by the dapp using CPI or by the owner of the account.
+It requires :
+
+* Account on which a fee was paid
+* Authority (owner) of the account (signer)
+
+Argument: Number of lamports to rebate (u64) can be u64::MAX to rebate all the fees.
+
+The authority or the owner could be easily deduced from the `AccountMeta`. In case of PDA's
+usually account and owner are the same (if it was not changed), then `invoke_signed` can be used
+to issue a rebate.
+In case of multiple rebate instructions, only the maximum rebate will one will be issued.
+Payer has to pay full application fees even if they are eligible for a rebate.
+If there is no application fee associated with the account we ignore the rebate instruction. 
+
+### Changes in the core solana code
+
+These are following changes that we have identified as required to implement this feature.
+
+#### Changes in `load_transaction_accounts`
+
+When we load transaction accounts we have to calculate the application fees by decoding the `PayApplicationFees`
+instruction. Then we verify that fee payer has miminum balance of:
```

Comment:
>  calculate the application fees by decoding the `PayApplicationFees` instruction.

Assume by now, `PayApplicationFee` IX has already been sanitized? (tho what sanitize entitles are yet to be defined)

Reply:
Usually we can only instruction that user can target externally is `PayApplicationFees` instruction. Other instructions are cpied into by the DApp.
So we can just check if the instruction targets application fee program and then decode it.

Does this answer the question or may be I misunderstood it.